File: backend/typescript/services/interfaces/userService.ts
import {
  CreateUserDTO,
  Role,
  SignUpMethod,
  UpdateUserDTO,
  UserDTO,
} from "../../types";

interface IUserService {
  /**
   * Retrieves a user by their ID.
   * @param userId The ID of the user to retrieve.
   * @returns A UserDTO containing the user's information or null if no user is found.
   * @throws Error if the retrieval process fails.
   */
  getUserById(userId: string): Promise<UserDTO>;

  /**
   * Retrieves a user by their email address.
   * @param email The email address of the user to retrieve.
   * @returns A UserDTO containing the user's information or null if no user is found.
   * @throws Error if the retrieval process fails.
   */
  getUserByEmail(email: string): Promise<UserDTO>;

  /**
   * Retrieves the role of a user by their authentication ID.
   * @param authId The authentication ID associated with the user.
   * @returns The role of the user or null if the user cannot be found.
   * @throws Error if the retrieval process fails.
   */
  getUserRoleByAuthId(authId: string): Promise<Role>;

  /**
   * Retrieves the ID of a user by their authentication ID.
   * @param authId The authentication ID associated with the user.
   * @returns The user's ID or null if the user cannot be found.
   * @throws Error if the retrieval process fails.
   */
  getUserIdByAuthId(authId: string): Promise<string>;

  /**
   * Retrieves the authentication ID associated with a user ID.
   * @param userId The user's ID to lookup the authentication ID.
   * @returns The authentication ID or null if the user cannot be found.
   * @throws Error if the retrieval process fails.
   */
  getAuthIdById(userId: string): Promise<string>;

  /**
   * Retrieves all users in the system.
   * @returns An array of UserDTOs, which could be empty if no users are found.
   * @throws Error if the retrieval process fails.
   */
  getUsers(): Promise<Array<UserDTO>>;

  /**
   * Creates a user with the specified details.
   * @param user The details of the user to create.
   * @param authId Optional; the user's Firebase authentication ID.
   * @param signUpMethod Optional; the method the user used to sign up.
   * @returns A UserDTO containing the created user's information.
   * @throws Error if the user creation process fails.
   */
  createUser(
    user: CreateUserDTO,
    authId?: string,
    signUpMethod?: SignUpMethod,
  ): Promise<UserDTO>;

  /**
   * Updates the information of a user by their ID.
   * @param userId The ID of the user to update.
   * @param user The new details to update the user with.
   * @returns A UserDTO containing the updated user's information.
   * @throws Error if the update process fails.
   */
  updateUserById(userId: string, user: UpdateUserDTO): Promise<UserDTO>;

  /**
   * Deletes a user by their user ID.
   * @param userId The ID of the user to delete.
   * @throws Error if the deletion process fails.
   */
  deleteUserById(userId: string): Promise<void>;

  /**
   * Deletes a user by their email address.
   * @param email The email of the user to delete.
   * @throws Error if the deletion process fails.
   */
  deleteUserByEmail(email: string): Promise<void>;
}

export default IUserService;

File: backend/python/app/services/interfaces/user_service.py
from abc import ABC, abstractmethod


class IUserService(ABC):
    """
    UserService interface with user management methods
    """

    @abstractmethod
    def get_user_by_id(self, user_id):
        """
        Get user associated with user_id

        :param user_id: user's id
        :type user_id: str
        :return: a UserDTO with user's information
        :rtype: UserDTO
        :raises Exception: if user retrieval fails
        """
        pass

    @abstractmethod
    def get_user_by_email(self, email):
        """
        Get user associated with email

        :param email: user's email
        :type email: str
        :return: a UserDTO with user's information
        :rtype: UserDTO
        :raises Exception: if user retrieval fails
        """
        pass

    @abstractmethod
    def get_user_role_by_auth_id(self, auth_id):
        """
        Get role of user associated with auth_id

        :param auth_id: user's auth_id
        :type auth_id: str
        :return: role of the user
        :rtype: str
        :raises Exception: if user role retrieval fails
        """
        pass

    @abstractmethod
    def get_user_id_by_auth_id(self, auth_id):
        """
        Get id of user associated with auth_id

        :param auth_id: user's auth_id
        :type auth_id: str
        :return: id of the user
        :rtype: str
        :raises Exception: if user_id retrieval fails
        """
        pass

    @abstractmethod
    def get_auth_id_by_user_id(self, user_id):
        """
        Get auth_id of user associated with user_id

        :param user_id: user's id
        :type user_id: str
        :return: auth_id of the user
        :rtype: str
        :raises Exception: if auth_id retrieval fails
        """
        pass

    @abstractmethod
    def get_users(self):
        """
        Get all users (possibly paginated in the future)

        :return: list of UserDTOs
        :rtype: [UserDTO]
        :raises Exception: if user retrieval fails
        """
        pass

    @abstractmethod
    def create_user(self, user, auth_id=None, signup_method="PASSWORD"):
        """
        Create a user, email verification configurable

        :param user: the user to be created
        :type user: CreateUserDTO
        :param auth_id: user's firebase auth id, defaults to None
        :type auth_id: string, optional
        :param signup_method: method of signup, either "PASSWORD" or "GOOGLE", defaults to "PASSWORD"
        :type signup_method: str, optional
        :return: the created user
        :rtype: UserDTO
        :raises Exception: if user creation fails
        """
        pass

    @abstractmethod
    def update_user_by_id(self, user_id, user):
        """
        Update a user
        Note: the password cannot be updated using this method, use IAuthService.reset_password instead

        :param user_id: user's id
        :type user_id: str
        :param user: the user to be updated
        :type user: UpdateUserDTO
        :return: the updated user
        :rtype: UserDTO
        :raises Exception: if user update fails
        """
        pass

    @abstractmethod
    def delete_user_by_id(self, user_id):
        """
        Delete a user by user_id

        :param user_id: user_id of user to be deleted
        :type user_id: str
        :raises Exception: if user deletion fails
        """
        pass

    @abstractmethod
    def delete_user_by_email(self, email):
        """
        Delete a user by email

        :param str email: email of user to be deleted
        :type email: str
        :raises Exception: if user deletion fails
        """
        pass
